### PR TITLE
Implementing Customizable Event Loop Groups and Domain Socket Channels for Netty Channel Configuration

### PR DESCRIPTION
**Pull request by SanHalacogluImproving**
Added ThreadPoolResources and ThreadPoolResourceAllocators to set up and select event loop groups and domain socket channel classes. This allows advanced users to use custom event loop groups and domain socket channels by extending ThreadPoolResource. This resource helps in configuring Netty channels. The ThreadPoolResourceAllocator is responsible for the default scenario, automatically selecting a platform-specific ThreadPoolResource when no custom configuration is provided.

### CODE REVIEW

File: java/client/src/main/java/glide/api/models/configuration/BaseClientConfiguration.java [hunk 1]
```diff
@@ -51,4 +52,13 @@ public abstract class BaseClientConfiguration {
    * it will result in a timeout error. If not set, a default value will be used.
    */
   private final Integer requestTimeout;
+
+  /**
+   * Field for customizing the Event Loop Group and Channel Configuration in Netty applications.
+   * Advanced users can utilize {@link glide.connectors.resources.KQueuePoolResource}/{@link
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/configuration/BaseClientConfiguration.java:57]: I agree with Jonathan.  Change it to something like... 
```
   * Advanced users can pass an extended {@link glide.connectors.resources. ThreadPoolResource} to pass  a user-defined event loop group.  Users are responsible for shutting the resource down when no longer in use.  
```

File: java/client/src/main/java/glide/connectors/resources/ThreadPoolResource.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package glide.connectors.resources;
+
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.unix.DomainSocketChannel;
+import lombok.Getter;
+import lombok.NonNull;
+
+/**
+ * Abstract base class that sets up the EventLoopGroup and channel configuration for Netty
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/connectors/resources/ThreadPoolResource.java:9]:
```suggestion
 * Abstract base class that provides the EventLoopGroup and DomainSocketChannel to be used by the Netty protocol. 
```

File: java/client/src/test/java/glide/managers/ConnectionManagerTest.java [hunk 3]
```diff
@@ -209,7 +209,8 @@ public void onConnection_emptyResponse_throwsClosingException() {
 
     assertTrue(executionException.getCause() instanceof ClosingException);
     assertEquals("Unexpected empty data in response", executionException.getCause().getMessage());
-    verify(channel).close();
+    // TODO: channel may not be completed since we don't wait for the close to complete
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/managers/ConnectionManagerTest.java:212]: this conflict will need to be fixed

File: java/client/src/main/java/glide/api/RedisClient.java [hunk 4]
```diff
@@ -21,21 +24,26 @@ public class RedisClient extends BaseClient implements BaseCommands {
    * Request an async (non-blocking) Redis client in Standalone mode.
    *
    * @param config - Redis Client Configuration
-   * @return a Future to connect and return a RedisClient
+   * @return a future to connect and return a RedisClient
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/RedisClient.java:27]: why?

**Review by acarbonetto — APPROVED** [review]

File: java/client/src/main/java/glide/api/models/configuration/BaseClientConfiguration.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/configuration/BaseClientConfiguration.java:57]: + you can shorten class ID by adding a corresponding import, so it will be like
```
... {@link KQueuePoolResource}/{@link EpollResource} ...
```

File: java/client/src/test/java/glide/api/RedisClientCreateTest.java [hunk 5]
```diff
@@ -78,8 +84,8 @@ public void createClient_errorOnConnectionThrowsExecutionException() {
     connectToRedisFuture.completeExceptionally(exception);
     RedisClientConfiguration config = RedisClientConfiguration.builder().build();
 
-    when(connectionManager.connectToRedis(eq(config))).thenReturn(connectToRedisFuture);
-    mockedClient.when(() -> CreateClient(config)).thenCallRealMethod();
+    when(connectionManager.connectToRedis(any())).thenReturn(connectToRedisFuture);
+    mockedClient.when(() -> CreateClient(any())).thenCallRealMethod();
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientCreateTest.java:88]: can we revert this change?  This would mean that we know which prototype message gets sent.

**Review by acarbonetto — APPROVED** [review]

**Review by jonathanl-bq — APPROVED** [review]
Approved, given that the last couple of comments get resolved.
